plugins/state_history_plugin/state_history_plugin.cpp

```diff
@@ -180,46 +180,34 @@ struct state_history_plugin_impl : std::enable_shared_from_this<state_history_pl
       }
 
       void start_read() {
-         auto in_buffer = std::make_shared<boost::beast::flat_buffer>();
-         derived_session().socket_stream->async_read(
-             *in_buffer, [self = derived_session().shared_from_this(), in_buffer](boost::system::error_code ec, size_t) {
-                self->callback(ec, "async_read", [self, in_buffer] {
-                   auto d = boost::asio::buffer_cast<char const*>(boost::beast::buffers_front(in_buffer->data()));
-                   auto s = boost::asio::buffer_size(in_buffer->data());
-                   fc::datastream<const char*> ds(d, s);
-                   state_request               req;
-                   fc::raw::unpack(ds, req);
-                   std::visit(*self, std::move(req));
-                });
-             });
+         // use reactor pattern with blocking read to avoid mixing async_read() and block write(). 
+         derived_session().socket_stream->next_layer().async_wait(
```

Comment:
Does this work correctly if two websocket payload frames are delivered "atomically" to the socket? My concern would be the async_wait() is fired once, the websocket::stream::read() reads in two payloads from the underlying socket, and then returns on the first payload while buffering the second one for the next read(). At that point the async_wait would not be fired for the second payload because the socket has been drained from the kernel's standpoint.